woocommerce/woocommerce-gateway-stripe
# Update order if there is an error while Stripe confirms payment

# Changes proposed in this Pull Request:

Issue: Link to the GitHub issue this PR addresses (if appropriate).
Fixes #1740 

Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.

This PR adds an AJAX function to handle errors that may occur while Stripe is confirming payment for an order. Previously we simply threw an error on the client and displayed a message to the user if `confirmPayment` or `confirmSetup` returned an error: however, we should be updating the order appropriately as well, in this scenario. This is predominantly catered towards payment methods without redirection (e.g. cards, SEPA); redirection payment methods (e.g. giropay) succeed or fail after authorisation and we handle these errors safely after the redirect. Until now we did not really have any error handling if an error appeared prior to this.

Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.


# Testing instructions

Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?

1. Please pull the latest changes from this PR and enable the UPE, of course.
2. Add an item to cart and complete a checkout using one of [these cards](https://stripe.com/docs/testing#cards-responses) that will elicit an error response.
3. Notice that an error is produced at checkout as usual. 
4. Go to your Orders inside WP-Admin and notice that an order has been created and that the status of the order is now Failed. 
5. Complete the order or continue producing errors and confirming that the status of the order is updated appropriately.
6. Test the above steps from the Pay for Order page and confirm the same results. 

-------------------

(remove me)
